### PR TITLE
Refactor Status class for hopefully increased legibility and reasoning

### PR DESCRIPTION
**Pull request by mjgiarlo**
I tested this PR by pointing [Argo at this branch](https://github.com/sul-dlss/argo/compare/main...workflow-client-v6), deploying Argo, and running the bulk metadata update integration test which we believe should test the code path changed here.

### CODE REVIEW

File: lib/dor/workflow/client/status.rb [hunk 1]
```diff
@@ -39,43 +39,12 @@ def initialize(druid:, version:, lifecycle_routes:)
...
-                next unless STEPS.key?(m_name) && (!status_time || m_time >= status_time)
-
-                status_code = STEPS[m_name]
-                status_time = m_time
-              end
-
-              { status_code: status_code, status_time: status_time }
-            end
-          end
-        end
```
Comment (mjgiarlo) [inline lib/dor/workflow/client/status.rb:68]: This was a public method, but as far as I can tell, it's only ever called within the class, and its purpose is to set two bits of state: a status code and a status time, which will both come from the the latest accessioned milestone or the latest current milestone (or default values if no current milestones). This commit dumps the method and sets the status code and status time values as part of instance initialization.

File: lib/dor/workflow/client/status.rb [hunk 2]
```diff
@@ -101,24 +82,33 @@ def simplified_status_code(display)
...
+          milestones
+            # milestone name must be in list of known steps
+            .select { |m| STEPS.key?(m[:milestone]) }
             # Two possible ways the version can indicate the milestone is part of the current version:
             # if m[:version] is nil, then the milestone is active (version 0 becoming version 1)
             # if m[:version] is matches the current version, then the milestone is archived with the current version
-            next unless m[:version].nil? || m[:version] == version
+            .select { |m| m[:version].nil? || m[:version] == version }
+            # registered milestone is only valid for v1
+            .reject { |m| m[:milestone] == 'registered' && version.to_i > 1 }
```
Comment (mjgiarlo) [inline lib/dor/workflow/client/status.rb:93]: Another focus of the remainder of this commit is on smarter use of Ruby iterator methods. I find that using `#each` can conceal intention, and this is a good example. We're using `#each` here as a way to filter a list into shorter list. To reveal that intention more clearly, refactor this as a set of selects and rejects.

File: lib/dor/workflow/client/status.rb [hunk 3]
```diff
@@ -39,43 +39,12 @@ def initialize(druid:, version:, lifecycle_routes:)
...
-            accessioned_milestones = current_milestones.select { |m| m[:milestone] == 'accessioned' }
-            if accessioned_milestones.any?
-              # if we have an accessioned milestone, this is the last possible step and should be the status regardless of timestamp
-              { status_code: STEPS['accessioned'], status_time: accessioned_milestones.last[:at].utc.xmlschema }
-            else
-              status_code = 0
-              status_time = nil
-              # for each milestone in the current version, see if it comes at the same time or after the current 'last' step.
-              # if so, make it the last and record the date/time
-              current_milestones.each do |m|
```
Comment (mjgiarlo) [inline lib/dor/workflow/client/status.rb:56]: And this is another instance of `#each` as an intention-befuddling method. Here the `#each` is acting as both a filter (removing milestones without a STEPS-matching name) and a max-value finder. But Ruby already has iterator methods that can do both of these operations.

File: lib/dor/workflow/client/status.rb [hunk 4]
```diff
@@ -101,24 +82,33 @@ def simplified_status_code(display)
...
+        end
 
-            current << m
-          end
-          current
+        def latest_accessioned_milestone
+          current_milestones
+            .select { |m| m[:milestone] == 'accessioned' }
+            .max_by { |m| m[:at].utc.xmlschema }
         end
```
Comment (mjgiarlo) [inline lib/dor/workflow/client/status.rb:108]: I added some small helper methods to give names to things.

**Comment by ndushay** [comment]
you may want to regen `.rubocop_todo.yml` as I expect some of the complexity numbers went down.

**Review by justinlittman — APPROVED** [review]

**Comment by mjgiarlo** [comment]
@ndushay 💬 
> you may want to regen `.rubocop_todo.yml` as I expect some of the complexity numbers went down.

Good call. I included that in the version bump commit I just pushed.

**Pull request merged**